src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,20 +27,86 @@ public EntityMappingImpl( final Graph mappingDescription ) {
-			// TODO ...
+			if(g2lMap.containsKey(it.next().getSubject())) {
+				// There is already at least one local URI for this global URI,
+				// We don't want to override it, so we'll add the new one to the set.
+				final Set<Node> locals = g2lMap.get(it.next().getSubject());
+				locals.add(it.next().getObject());
+				g2lMap.replace(it.next().getSubject(),locals);
+			} else {
+				// g2lmap doesn't contain any local URI for this global yet
+				final Set<Node> locals = new HashSet<Node>();
+				locals.add(it.next().getObject());
+				g2lMap.put(it.next().getSubject(),locals);
+			}
+			
+			// The same for l2g.
+			if(l2gMap.containsKey(it.next().getObject())) {
+				final Set<Node> globals = l2gMap.get(it.next().getObject());
+				globals.add(it.next().getSubject());
+				l2gMap.replace(it.next().getObject(),globals);
+			} else {
+				final Set<Node> globals = new HashSet<Node>();
+				globals.add(it.next().getSubject());
+				l2gMap.put(it.next().getObject(),globals);
+			}
 		}
 	}
 
 	@Override
 	public Set<TriplePattern> applyToTriplePattern( final TriplePattern tp ) {
-		// TODO: implement this function
-		return Collections.singleton(tp);
+		final Set<Triple> workingSet = new HashSet<Triple>();
+		if(g2lMap.containsKey(tp.asJenaTriple().getSubject())) {
+			final Iterator<Node> it = g2lMap.get(tp.asJenaTriple().getSubject()).iterator();
+			while (it.hasNext()) {
+				final Triple workingTriple = new Triple(it.next(),tp.asJenaTriple().getPredicate(),tp.asJenaTriple().getObject());
+				workingSet.add(workingTriple);
+			}
+		} else {
+			workingSet.add(tp.asJenaTriple());
+		}
+
+		final Set<Triple> workingSet2 = new HashSet<Triple>();
+		if(g2lMap.containsKey(tp.asJenaTriple().getPredicate())) {
+			final Iterator<Triple> it0 = workingSet.iterator();
+			while (it0.hasNext()) {
+				final Triple workingTriple0 = it0.next();
+				final Iterator<Node> it = g2lMap.get(tp.asJenaTriple().getPredicate()).iterator();
+				while (it.hasNext()) {
+					final Triple workingTriple = new Triple(workingTriple0.getSubject(),it.next(),workingTriple0.getObject());
+					workingSet2.add(workingTriple);
+				}
+			}
+		} else {
+			workingSet2.addAll(workingSet);
+		}
+		
+
+		final Set<TriplePattern> resultSet = new HashSet<TriplePattern>();
+		if(g2lMap.containsKey(tp.asJenaTriple().getObject())) {
+			final Iterator<Triple> it0 = workingSet2.iterator();
+			while (it0.hasNext()) {
+				final Triple workingTriple0 = it0.next();
+				final Iterator<Node> it = g2lMap.get(tp.asJenaTriple().getObject()).iterator();
+				while (it.hasNext()) {
+					final Triple workingTriple = new Triple(workingTriple0.getSubject(),workingTriple0.getPredicate(),it.next());
+					resultSet.add(new TriplePatternImpl(workingTriple));
+				}
+			}
+			return resultSet;
+		} else {
+			final Iterator<Triple> it = workingSet2.iterator();
+			while (it.hasNext()) {
+				resultSet.add(new TriplePatternImpl(it.next()));
+			}
+			return resultSet;
+		}
```

Comment:
All the comments from above apply here too (check first that the object of `tp` is a URI, avoid unnecessary lookups into `g2lMap`, instead of iterators better use for-loops).